### PR TITLE
deploy endpoint 🔥 

### PR DESCRIPTION
**Pull request by jkminneti**
🔥 🚒 🧯

### CODE REVIEW

File: cmd/rollback.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	Use:           "rollback [app]",
+	Short:         "immediately roll back an app",
+	RunE:          runRollback,
+	PreRunE:       promptCurrentContext,
+	Long: `immediately rolls back to the resources (and image) applied during the last successful release, without monitoring for success.
+Can be used to abort a running release as well, as tuber's definition of 'last successful release' is not updated until a running release finishes successfully.`,
+}
+
+func runRollback(cmd *cobra.Command, args []string) error {
+	appName := args[0]
```
Comment (jaysonesmith) [inline cmd/rollback.go:23]: Same C2 feedback as with `deploy`.

**Review by jaysonesmith — CHANGES_REQUESTED** [review]
Great work! Just a couple checks needed.

File: graph/schema.resolvers.go [hunk 2]
```diff
@@ -70,6 +72,36 @@ func (r *mutationResolver) RemoveApp(ctx context.Context, input model.AppInput)
 	return app, nil
 }
 
+func (r *mutationResolver) Deploy(ctx context.Context, input model.DeployInput) (*model.TuberApp, error) {
```
[the remark this answers is not in the document]
Comment (jkminneti) [reply graph/schema.resolvers.go:77]: autogenerated :(

File: cmd/deploy.go [hunk 3]
```diff
@@ -2,28 +2,63 @@ package cmd
...
-	Use:          "deploy [app]",
-	Short:        "deploys the latest built image of an app. CURRENTLY REQUIRES A LOCAL DB",
-	RunE:         deploy,
-	PreRunE:      promptCurrentContext,
+	SilenceErrors: true,
+	SilenceUsage:  true,
+	Use:           "deploy [app] -t [tag]",
+	Short:         "deploys the latest built image of an app, or a certain tag if specified",
+	RunE:          deploy,
+	PreRunE:       promptCurrentContext,
```
[the remark this answers is not in the document]
Comment (jkminneti) [reply cmd/deploy.go:21]:
```go
Args: cobra.ExactArgs(1),
```
but will do good catch

[the remark this answers is not in the document]
Comment (jaysonesmith) [reply cmd/deploy.go:21]: The issue with `ExactArgs` is that when you've got `SilenceUsage: true` set, all you get is, "2021/06/14 15:44:15 accepts 1 arg(s), received 0" which is super unhelpful.

File: cmd/deploy.go [hunk 4]
```diff
@@ -2,28 +2,63 @@ package cmd
...
 
 var deployCmd = &cobra.Command{
-	SilenceUsage: true,
-	Use:          "deploy [app]",
-	Short:        "deploys the latest built image of an app. CURRENTLY REQUIRES A LOCAL DB",
-	RunE:         deploy,
-	PreRunE:      promptCurrentContext,
+	SilenceErrors: true,
+	SilenceUsage:  true,
+	Use:           "deploy [app] -t [tag]",
```
[the remark this answers is not in the document]
Comment (jkminneti) [reply cmd/deploy.go:17]: wow we have done that wrong literally everywhere. good good call

File: cmd/deploy.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jkminneti) [reply cmd/deploy.go:20]: ugh that's frustrating. but i'd question if `"app name is required"` is sufficient either

that could be a flag or something who knows

so i _want_ cobra to only display usage whenever Args errors, but it doesnt :/

[the remark this answers is not in the document]
Comment (jkminneti) [reply cmd/deploy.go:20]: given that this affects basically every command in the book though, i dont want to fix this now, and would rather come up with a global solution separately

[the remark this answers is not in the document]
Comment (jkminneti) [reply cmd/deploy.go:20]: to be clear i did leave it with ExactArgs for now

**Review by quinn-freshly — APPROVED** [review]

**Pull request merged**